hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java

```diff
@@ -849,6 +858,40 @@ private void validateSnapshotsExistAndActive(final String volumeName,
     checkSnapshotActive(toSnapInfo, false);
   }
 
+  public void snapshotLimitCheck() throws IOException, OMException {
+    OmMetadataManagerImpl omMetadataManager = (OmMetadataManagerImpl) ozoneManager.getMetadataManager();
+    SnapshotChainManager snapshotChainManager = omMetadataManager.getSnapshotChainManager();
+    int currentSnapshotNum = snapshotChainManager.getGlobalSnapshotChain().size();
+    int oldCount = inFlightSnapshotCount.get();
+    int newCount = inFlightSnapshotCount.updateAndGet(count -> {
```

Comment:
Since we are using two steps, get and update, we may throw an exception for a few snapshot requests due to the case condition between get and update. It is not a big deal, but we can avoid it by putting it inside a synchronized block on `inFlightSnapshotCount` object.

Reply:
agreed this section is prone to race conditions.

Semaphore would be a more appropriate solution, but it will be require a larger rewrite.

Reply:
Why can't we directly have an AtomicReference<OMException> outside the block? This can be just set to the exception that needs to be thrown if at all it needs to be thrown
with a null check it can be checked outside of the getAndUpdate

Reply:
We wouldn't need the oldCount and newCount with the above proposed change